File: smart_monitor.py
# smart_monitor.py — Python версия

import subprocess
import json
import re
import sys
import os
import csv
from datetime import datetime
from colorama import init, Fore, Style

init(autoreset=True)

class SMARTMonitor:
    def __init__(self):
        self.devices = []
        self.smart_data = {}

    def get_devices(self):
        """Получает список всех дисков с помощью smartctl --scan."""
        try:
            output = subprocess.check_output(['smartctl', '--scan'], 
                                              stderr=subprocess.DEVNULL,
                                              universal_newlines=True)
            for line in output.strip().split('\n'):
                if line:
                    parts = line.split()
                    if parts:
                        self.devices.append(parts[0])
        except FileNotFoundError:
            print(Fore.RED + "❌ smartctl не найден. Установите smartmontools.")
            sys.exit(1)
        return self.devices

    def get_smart_data(self, device):
        """Получает SMART-данные для устройства в формате JSON."""
        try:
            output = subprocess.check_output(
                ['smartctl', '-j', '-a', device],
                stderr=subprocess.DEVNULL,
                universal_newlines=True
            )
            return json.loads(output)
        except subprocess.CalledProcessError:
            return None
        except json.JSONDecodeError:
            return None

    def parse_attributes(self, data):
        """Извлекает ключевые SMART-атрибуты."""
        if not data:
            return None

        result = {
            'model': data.get('model_name', 'Unknown'),
            'serial': data.get('serial_number', 'Unknown'),
            'firmware': data.get('firmware_version', 'Unknown'),
            'smart_supported': data.get('smart_supported', False),
            'smart_enabled': data.get('smart_enabled', False),
            'smart_status': data.get('smart_status', {}).get('passed', False),
            'temperature': None,
            'power_on_hours': None,
            'power_cycles': None,
            'read_errors': None,
            'write_errors': None,
            'reallocated_sectors': None,
            'pending_sectors': None,
            'uncorrectable_sectors': None,
        }

        # Извлекаем атрибуты из ata_smart_attributes или nvme_smart_health_information_log
        if 'ata_smart_attributes' in data:
            attrs = data['ata_smart_attributes'].get('table', [])
            for attr in attrs:
                name = attr.get('name', '')
                value = attr.get('value', 0)
                raw = attr.get('raw', {}).get('value', 0)
                if name == 'Temperature_Celsius':
                    result['temperature'] = raw
                elif name == 'Power_On_Hours':
                    result['power_on_hours'] = raw
                elif name == 'Power_Cycle_Count':
                    result['power_cycles'] = raw
                elif name == 'Read_Error_Rate':
                    result['read_errors'] = raw
                elif name == 'Write_Error_Rate':
                    result['write_errors'] = raw
                elif name == 'Reallocated_Sector_Ct':
                    result['reallocated_sectors'] = raw
                elif name == 'Current_Pending_Sector':
                    result['pending_sectors'] = raw
                elif name == 'Offline_Uncorrectable':
                    result['uncorrectable_sectors'] = raw

        # NVMe поддержка
        if 'nvme_smart_health_information_log' in data:
            nvme = data['nvme_smart_health_information_log']
            result['temperature'] = nvme.get('temperature', result['temperature'])
            result['power_on_hours'] = nvme.get('power_on_hours', result['power_on_hours'])
            result['power_cycles'] = nvme.get('power_cycles', result['power_cycles'])

        return result

    def scan(self):
        """Сканирует все диски и собирает SMART-данные."""
        self.get_devices()
        for device in self.devices:
            print(f"🔍 Проверка {device}...")
            data = self.get_smart_data(device)
            if data:
                parsed = self.parse_attributes(data)
                if parsed:
                    self.smart_data[device] = parsed

    def print_table(self):
        """Выводит таблицу с данными."""
        if not self.smart_data:
            print(Fore.YELLOW + "Нет данных для отображения.")
            return

        print(Fore.CYAN + "\n📊 Состояние дисков:")
        print("┌" + "─" * 100 + "┐")
        print(f"│ {'Диск':<12} {'Модель':<20} {'Темп.':<8} {'Время':<10} "
              f"{'Ошибки':<8} {'Realloc':<8} {'Состояние':<12} │")
        print("├" + "─" * 100 + "┤")

        for device, data in self.smart_data.items():
            temp = f"{data.get('temperature', 'N/A')}°C" if data.get('temperature') else 'N/A'
            hours = f"{data.get('power_on_hours', 0)}h" if data.get('power_on_hours') else 'N/A'
            errors = data.get('read_errors', 0) or 0
            realloc = data.get('reallocated_sectors', 0) or 0
            status = data.get('smart_status', False)

            status_color = Fore.GREEN if status else Fore.RED
            status_text = "✅ OK" if status else "❌ FAIL"

            # Цвет для температуры
            temp_color = Fore.GREEN
            if data.get('temperature') and data['temperature'] > 50:
                temp_color = Fore.RED
            elif data.get('temperature') and data['temperature'] > 40:
                temp_color = Fore.YELLOW

            model = data.get('model', 'Unknown')[:20]
            print(f"│ {device:<12} {model:<20} {temp_color}{temp:<8}{Style.RESET_ALL} "
                  f"{hours:<10} {errors:<8} {realloc:<8} {status_color}{status_text:<12}{Style.RESET_ALL} │")

        print("└" + "─" * 100 + "┘")

    def save_json(self, filename='smart_report.json'):
        with open(filename, 'w', encoding='utf-8') as f:
            json.dump(self.smart_data, f, indent=2, ensure_ascii=False)
        print(Fore.GREEN + f"💾 Сохранено JSON: {filename}")

    def save_csv(self, filename='smart_report.csv'):
        if not self.smart_data:
            return
        with open(filename, 'w', newline='', encoding='utf-8') as f:
            writer = csv.writer(f)
            writer.writerow(['Device', 'Model', 'Serial', 'Temperature', 'PowerOnHours',
                             'ReadErrors', 'WriteErrors', 'ReallocatedSectors',
                             'PendingSectors', 'SMARTStatus'])
            for device, data in self.smart_data.items():
                writer.writerow([
                    device,
                    data.get('model', ''),
                    data.get('serial', ''),
                    data.get('temperature', ''),
                    data.get('power_on_hours', ''),
                    data.get('read_errors', ''),
                    data.get('write_errors', ''),
                    data.get('reallocated_sectors', ''),
                    data.get('pending_sectors', ''),
                    data.get('smart_status', False)
                ])
        print(Fore.GREEN + f"💾 Сохранено CSV: {filename}")

def main():
    print(Fore.CYAN + "💾 SMART Disk Monitor (Python)")
    monitor = SMARTMonitor()
    monitor.scan()
    monitor.print_table()

    if monitor.smart_data:
        monitor.save_json()
        monitor.save_csv()

if __name__ == "__main__":
    main()

File: smart_monitor.js
// smart_monitor.js — JavaScript версия

const { exec } = require('child_process');
const fs = require('fs');
const util = require('util');

const execPromise = util.promisify(exec);

class SMARTMonitor {
    constructor() {
        this.devices = [];
        this.smartData = {};
    }

    async getDevices() {
        try {
            const { stdout } = await execPromise('smartctl --scan');
            const lines = stdout.trim().split('\n');
            for (const line of lines) {
                if (line) {
                    const parts = line.split(/\s+/);
                    if (parts.length > 0) {
                        this.devices.push(parts[0]);
                    }
                }
            }
        } catch (error) {
            console.error('\x1b[31m❌ smartctl не найден. Установите smartmontools.\x1b[0m');
            process.exit(1);
        }
        return this.devices;
    }

    async getSmartData(device) {
        try {
            const { stdout } = await execPromise(`smartctl -j -a ${device}`);
            return JSON.parse(stdout);
        } catch (error) {
            return null;
        }
    }

    parseAttributes(data) {
        if (!data) return null;

        const result = {
            model: data.model_name || 'Unknown',
            serial: data.serial_number || 'Unknown',
            firmware: data.firmware_version || 'Unknown',
            smart_supported: data.smart_supported || false,
            smart_enabled: data.smart_enabled || false,
            smart_status: data.smart_status?.passed || false,
            temperature: null,
            power_on_hours: null,
            power_cycles: null,
            read_errors: null,
            write_errors: null,
            reallocated_sectors: null,
            pending_sectors: null,
            uncorrectable_sectors: null,
        };

        if (data.ata_smart_attributes) {
            const attrs = data.ata_smart_attributes.table || [];
            for (const attr of attrs) {
                const name = attr.name || '';
                const raw = attr.raw?.value || 0;
                if (name === 'Temperature_Celsius') result.temperature = raw;
                else if (name === 'Power_On_Hours') result.power_on_hours = raw;
                else if (name === 'Power_Cycle_Count') result.power_cycles = raw;
                else if (name === 'Read_Error_Rate') result.read_errors = raw;
                else if (name === 'Write_Error_Rate') result.write_errors = raw;
                else if (name === 'Reallocated_Sector_Ct') result.reallocated_sectors = raw;
                else if (name === 'Current_Pending_Sector') result.pending_sectors = raw;
                else if (name === 'Offline_Uncorrectable') result.uncorrectable_sectors = raw;
            }
        }

        if (data.nvme_smart_health_information_log) {
            const nvme = data.nvme_smart_health_information_log;
            result.temperature = nvme.temperature || result.temperature;
            result.power_on_hours = nvme.power_on_hours || result.power_on_hours;
            result.power_cycles = nvme.power_cycles || result.power_cycles;
        }

        return result;
    }

    async scan() {
        await this.getDevices();
        for (const device of this.devices) {
            console.log(`🔍 Проверка ${device}...`);
            const data = await this.getSmartData(device);
            if (data) {
                const parsed = this.parseAttributes(data);
                if (parsed) {
                    this.smartData[device] = parsed;
                }
            }
        }
    }

    printTable() {
        const data = this.smartData;
        if (Object.keys(data).length === 0) {
            console.log('\x1b[33mНет данных для отображения.\x1b[0m');
            return;
        }

        console.log('\x1b[36m\n📊 Состояние дисков:\x1b[0m');
        console.log('┌' + '─'.repeat(100) + '┐');
        console.log(`│ ${'Диск'.padEnd(12)} ${'Модель'.padEnd(20)} ${'Темп.'.padEnd(8)} ${'Время'.padEnd(10)} ${'Ошибки'.padEnd(8)} ${'Realloc'.padEnd(8)} ${'Состояние'.padEnd(12)} │`);
        console.log('├' + '─'.repeat(100) + '┤');

        for (const [device, info] of Object.entries(data)) {
            const temp = info.temperature ? `${info.temperature}°C` : 'N/A';
            const hours = info.power_on_hours ? `${info.power_on_hours}h` : 'N/A';
            const errors = info.read_errors || 0;
            const realloc = info.reallocated_sectors || 0;
            const status = info.smart_status ? '\x1b[32m✅ OK\x1b[0m' : '\x1b[31m❌ FAIL\x1b[0m';

            let tempColor = '\x1b[32m';
            if (info.temperature > 50) tempColor = '\x1b[31m';
            else if (info.temperature > 40) tempColor = '\x1b[33m';

            const model = (info.model || 'Unknown').slice(0, 20);
            console.log(`│ ${device.padEnd(12)} ${model.padEnd(20)} ${tempColor}${temp.padEnd(8)}\x1b[0m ${hours.padEnd(10)} ${String(errors).padEnd(8)} ${String(realloc).padEnd(8)} ${status.padEnd(12)} │`);
        }
        console.log('└' + '─'.repeat(100) + '┘');
    }

    saveJSON(filename = 'smart_report.json') {
        fs.writeFileSync(filename, JSON.stringify(this.smartData, null, 2));
        console.log(`\x1b[32m💾 Сохранено JSON: ${filename}\x1b[0m`);
    }

    saveCSV(filename = 'smart_report.csv') {
        const data = this.smartData;
        if (Object.keys(data).length === 0) return;
        let csv = 'Device,Model,Serial,Temperature,PowerOnHours,ReadErrors,WriteErrors,ReallocatedSectors,PendingSectors,SMARTStatus\n';
        for (const [device, info] of Object.entries(data)) {
            csv += `${device},${info.model || ''},${info.serial || ''},${info.temperature || ''},${info.power_on_hours || ''},${info.read_errors || ''},${info.write_errors || ''},${info.reallocated_sectors || ''},${info.pending_sectors || ''},${info.smart_status || false}\n`;
        }
        fs.writeFileSync(filename, csv);
        console.log(`\x1b[32m💾 Сохранено CSV: ${filename}\x1b[0m`);
    }
}

async function main() {
    console.log('\x1b[36m💾 SMART Disk Monitor (JavaScript)\x1b[0m');
    const monitor = new SMARTMonitor();
    await monitor.scan();
    monitor.printTable();

    if (Object.keys(monitor.smartData).length > 0) {
        monitor.saveJSON();
        monitor.saveCSV();
    }
}

main().catch(console.error);
